Update `ImageResolver` cache structure to use repo name as key

erikayasuda opened the pull request:
### What does this PR do?
Restructures the `ImageResolver` cache structure so that the keys are only the repository names (ex - `dd-lib-python-init`) and not the repository URL (ex - `gcr.io/dd-lib-python-init`). 

It also enables us to configure the default Datadog container registries via configuration. This is not public-facing, and will be used primarily for dogfooding the new feature on staging (given staging uses different container registries).

### Motivation
The repository URL field in the `K8S_INJECTION_DD` remote config data was not designed to consumed by DCA, but primarily for use by our internal `equilibrium` tooling. This meant that by using the repository URL for the key, it would be limited to only allowing gradual rollout for customers using `grc.io/datadoghq` as their registry. See [here](https://docs.datadoghq.com/tracing/trace_collection/automatic_instrumentation/single-step-apm/kubernetes/?tab=agentv764recommended#change-the-default-image-registry) for other valid Datadog registries.

### Describe how you validated your changes
- Updated existing tests
- Added new unit tests
- Local E2E testing with `injector-dev`

### Additional Notes

erikayasuda committed:
Updated ImageResolver cache structure to make repo name the key

erikayasuda committed:
Fix release note for gradual rollout

erikayasuda committed:
Update ImageResolver to do explicit registry checks to match updated cache format

erikayasuda committed:
Updated ResolvedImage to store ImageRef with no registry prepended, so it can be fully registry agnostic

erikayasuda commented on pkg/clusteragent/admission/mutate/autoinstrumentation/image_resolver.go:
```diff
@@ -116,38 +122,46 @@ func (r *remoteConfigImageResolver) waitForInitialConfig() error {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
 
+	if !isDatadoghqRegistry(registry) {
```
I think we can keep it here for now (but will move to before the mutex lock), since the registry string is already being passed in here anyways. I agree that it could probably happen earlier since the registry should stay constant, unless the deployment gets updated. I'm down to make that change for 7.72 unless we foresee issues with this.

erikayasuda committed:
Merge branch 'main' into erikayasuda/fix-cache-structure

erikayasuda committed:
Merge branch 'main' into erikayasuda/fix-cache-structure

erikayasuda committed:
Added ResolvedImage typing that would provide the fully resolved image ref with registry, repo name, and digest

erikayasuda committed:
Updated release note to specifically call out which registries are supported

erikayasuda committed:
Remove unnecessary test case

erikayasuda committed:
Fixed registry URLs

erikayasuda committed:
Fixed tests for DD registry checking

adel121 reviewed (commented):
LGTM

Left a minor suggestion.

erikayasuda committed:
Configurable DD default registries (#40855)

Enabling the default DD registries to be configurable to enable testing on staging

erikayasuda committed:
Merge branch 'main' into erikayasuda/fix-cache-structure
